### PR TITLE
[frenemies] Adds `prove_participation` method

### PR DESCRIPTION
**Pull request by damirka**
For participants to prove ownership of an account in the leaderboard, we add a simple method with tracking via events.

### CODE REVIEW

File: sui_programmability/examples/frenemies/sources/frenemies.move [hunk 1]
```diff
@@ -39,8 +39,15 @@ module frenemies::frenemies {
         epoch_score: u16,
     }
 
+    /// Emitted in `confirm_address` call.
+    struct ParticipationProofCommitted has copy, drop {
+        commitment_hash: vector<u8>
```
Comment (kchalkias) [inline sui_programmability/examples/frenemies/sources/frenemies.move:44]: Typically to prove account ownership we sign over a proof of possession challenge (imagine a personal message over some challenge string). 
Other ways are to submit a tx which includes as metadata something that the off chain challenger asked you.

Just committing to a hash onchain,  might not suffice to prove ownership offchain. The reason being you will reveal the pre image to the challenger (off chain verifier), but now this verifier can use your preimage to impersonate you in another challenger. 

TL;DR Typically proof of possession should protect against replay attacks, which means you require some unpredictable non-reusable freshness per challenge (proof), and thus you either do some ZKP or sign over a fresh nonce challenge each time you are requested to prove ownership.

Comment (damirka) [reply sui_programmability/examples/frenemies/sources/frenemies.move:44]: Also it’s a last minute change before code goes live.

Comment (damirka) [reply sui_programmability/examples/frenemies/sources/frenemies.move:44]: Using hash here to make sure people don’t submit a tx with sensitive data (say email or a discord login).

Comment (kchalkias) [reply sui_programmability/examples/frenemies/sources/frenemies.move:44]: Then don't name it commitment_hash, but rather nonce?

File: sui_programmability/examples/frenemies/sources/frenemies.move [hunk 2]
```diff
@@ -96,6 +103,18 @@ module frenemies::frenemies {
...
     }
 
+    /// To prove participation offchain, participants need to hash some information and
+    /// call this function to prove that they own the address in the leaderboard.
+    ///
+    /// Aborts if the value submitted is not a hash (based on the length - 32 bytes).
+    /// Dummy check to prevent unnecessary spam and mistakes on the caller end.
+    public entry fun prove_participation(commitment_hash: vector<u8>) {
+        assert!(std::vector::length(&commitment_hash) == 32, EHashExpected);
+        event::emit(ParticipationProofCommitted {
```
Comment (sblackshear) [inline sui_programmability/examples/frenemies/sources/frenemies.move:113]: Should we include the sender address in the event to make it easier to check this based on the event stream alone?

**Review by sblackshear — APPROVED** [review]

**Review by sblackshear — CHANGES_REQUESTED** [review]
On second thought, let's not rush this in--it can easily go in a different package, and there are some details to be worked out here.

**Comment by arty-arty** [comment]
Hello, guys :) 

Would there be no use for this? It's a proof of NFT possession, I developed some time ago.

With protection against replay attacks. And though you do not need it here with ability to trade NFT freely and still prove possession.
